crates/api/src/site.rs

```diff
@@ -547,3 +558,77 @@ impl Perform for SaveSiteConfig {
     Ok(GetSiteConfigResponse { config_hjson })
   }
 }
+
+#[async_trait::async_trait(?Send)]
+impl Perform for PurgeItem {
+  type Response = PurgeItemResponse;
+
+  async fn perform(
+    &self,
+    context: &Data<LemmyContext>,
+    _websocket_id: Option<ConnectionId>,
+  ) -> Result<Self::Response, LemmyError> {
+    let data: &Self = self;
+    let local_user_view =
+      get_local_user_view_from_jwt(&data.auth, context.pool(), context.secret()).await?;
+
+    // Only let admins purge an item
+    is_admin(&local_user_view)?;
+
+    if let Some(comment_id) = data.comment_id {
+      blocking(context.pool(), move |conn| {
+        Comment::delete(conn, comment_id)
+      })
+      .await??;
+    } else if let Some(post_id) = data.post_id {
+      blocking(context.pool(), move |conn| Post::delete(conn, post_id)).await??;
+    } else if let Some(person_id) = data.person_id {
+      if data.remove_images {
+        // Read the person to get their images
+        let person = blocking(context.pool(), move |conn| Person::read(conn, person_id)).await??;
+
+        if let Some(banner) = person.banner {
+          purge_image_from_pictrs(context.client(), &context.settings(), &banner.into_inner())
+            .await?;
+        }
+
+        if let Some(avatar) = person.avatar {
+          purge_image_from_pictrs(context.client(), &context.settings(), &avatar.into_inner())
+            .await?;
+        }
+      }
+
+      blocking(context.pool(), move |conn| Person::delete(conn, person_id)).await??;
```

Comment:
I think it would be much cleaner to move this logic into a trait method which gets implemented on all the types. Also because this needs to be done from apub code so that purges can be federated.

Reply:
It is a trait method, the `Crud` trait.

Reply:
Sorry I meant the whole block, everything inside the if.

Reply:
I spose I could, but the trait would need to be in this crate, because it needs to do both DB actions (in the DB crates), and pictrs removes (in utils crate).

So maybe `trait Purge`, `impl PurgeData for PersonId ...`

Reply:
Actually it looks like the DB queries crate already includes utils, so I could put it there.

Reply:
This doesn't really apply anymore since there are now 4 endpoints.